packages/next/lib/page-loader.js

```diff
@@ -110,10 +123,30 @@ export default class PageLoader {
     }
   }
 
+  async prefetch (route) {
+    route = this.normalizeRoute(route)
+    const scriptRoute = route === '/' ? '/index.js' : `${route}.js`
+    if (this.prefetchCache.has(scriptRoute)) {
+      return
+    }
+    this.prefetchCache.add(scriptRoute)
+
+    const link = document.createElement('link')
+    // Feature detection is used to see if prefetch is supported, else fall back to preload
+    // Mainly this is for Safari
+    // https://caniuse.com/#feat=link-rel-prefetch
+    // https://caniuse.com/#feat=link-rel-preload
+    link.rel = supportsPrefetch ? 'prefetch' : 'preload'
+    link.href = `${this.assetPrefix}/_next/static/${encodeURIComponent(this.buildId)}/pages${scriptRoute}`
+    link.as = 'script'
+    document.head.appendChild(link)
+  }
+
   clearCache (route) {
     route = this.normalizeRoute(route)
     delete this.pageCache[route]
     delete this.loadingRoutes[route]
+    delete this.loadingRoutes[route]
```

Comment:
@timneutkens this one looks duplicated, where you trying to delete from `prefetchCache`?

Reply:
Yeah I saw that while adding Malte's feedback, will be in my upcoming PR.